torch/utils/_sympy/functions.py

```diff
@@ -1,43 +1,78 @@
+import functools
 import math
+import sys
 
 import sympy
 from sympy import S
-from sympy.core.logic import fuzzy_and, fuzzy_not, fuzzy_or
 
 __all__ = [
     "FloorDiv",
     "ModularIndexing",
     "CleanDiv",
     "CeilDiv",
-    "Pow",
-    "TrueDiv",
+    "IntTrueDiv",
+    "FloatTrueDiv",
     "LShift",
     "RShift",
     "IsNonOverlappingAndDenseIndicator",
-    "Round",
+    "RoundToInt",
     "RoundDecimal",
+    "ToFloat",
+    "FloatPow",
+    "PowByNatural",
 ]
 
 
+def _keep_float(f):
+    @functools.wraps(f)
+    def inner(*args):
+        r = f(*args)
+        if any(isinstance(a, sympy.Float) for a in args) and not isinstance(
+            r, sympy.Float
+        ):
+            r = sympy.Float(float(r))
+        return r
+
+    return inner
+
+
 def fuzzy_eq(x, y):
     if None in (x, y):
         return None
     return x == y
 
 
+# It would be nice to have assertions on whether or not inputs is_integer
+# However, with bugs like https://github.com/sympy/sympy/issues/26620 sympy
+# sometimes inconsistently reports floats an integers.
+#
+# What we can assume from sympy is that if something is an int, it
+# definitely is is_integer, but if it is a float it may or may not
+# be is_integer.  So we are unable to do strong asserts that things
+# are NOT integers.
+
+
+# TODO: In Triton, // rounds to zero, but in Python, it is floor division.
+# When we can prove both arguments are non-negative, we should just have a
+# GenericFloorDiv (name pending) which can codegen efficiently in Python/C,
```

Comment:
I feel it is important to know if the function is guaranteed to be operating only on non-negative inputs at the Sympy level. I don't think it mattered for modulus, but here is an example from FloorDiv which I have been working on:

```
        # Note [Integer division of fractions]
        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        #
        # This formula (a / b) / c == a / (b * c) is valid only for integer
        # division when the arguments are positive, or when the division is
        # Euclidean division or C-style truncation division.
        #
        #   (declare-const a Int)
        #   (declare-const b Int)
        #   (declare-const c Int)
        #   (assert (>= a 0))
        #   (assert (> b 0))
        #   (assert (> c 0))
        #   (assert (not (= (div (div a b) c) (div a (* b c)))))
        #   (check-sat)
        #   (get-model)
        #
        # is unsat in Z3, as is this formula with truncdiv:
        #
        #   (define-fun truncdiv ((x Int) (y Int)) Int
        #     (ite (or (= (mod x y) 0) (>= x 0))
        #          (div x y)
        #          (ite (>= y 0)
        #               (+ (div x y) 1)
        #               (- (div x y) 1))))
        #
        # (h/t https://gist.github.com/Rufflewind/a880e03fb0d13644a1e8)
        #
        # Counterexample for floor division when arguments are not positive: a=11, b=-3, c=-4
        #
        # Found using:
        #   (define-fun floor_div ((x Int) (y Int)) Int
        #     (to_int (/ (to_real x) (to_real y))))
        #
        #   >>> (11 // -3) // -4
        #   1
        #   >>> 11 // (-3 * -4)
        #   0
        #
        # Counterexample for Euclidean division: a=71, b=-9, c=-6
        #
        #   >>> euclidean_div(euclidean_div(71, -9), -6)
        #   2
        #   >>> euclidean_div(71, -9 * -6)
        #   1
        #
```

If we represent floor divisions from Python source code as PythonFloorDiv, we can no longer perform this simplification rule at the Sympy level.

Reply:
Re whether or not we should do it with a global shape env in Inductor, well, Inductor already has a global shape env, so assuming you are in Inductor code you always have the capability to query it. I think what I least understand how to do is how to deal with this situation: Inductor will often allocate fresh sympy variables representing indexing things that aren't tracked by ShapeEnv, so at minimum we would need to augment ShapeEnv with the ability to deal with these things. But let's say I'm doing codegen for `floordiv` and I want to know if I can output a `//`. How do I query if the LHS and RHS are positive? These are arbitrary sympy expressions that possibly depend on indexing variables like x0/i0, how do I know what the initial bounds for these variables are? It seems difficult to know contextually what I should have done here.

Reply:
I am also not sure what you don't like about having a partial function which is UB when the inputs are negative. This is a relatively common thing to do in compilers, arguably, this is exactly the sort of situation for which UB was invented for in the first place.

Reply:
For how to query whether the LHS and RHS are positive, see the changes in `index_propagation.py` in https://github.com/pytorch/pytorch/pull/114471.

My point here is not that we can't have more classes to work around these limitations, my point is that, at the moment, we are repeating this pattern of "I need to simplify something, but SymPy does not know enough as to simplify it itself, so let me add extra code to simplify it myself". See for example the `simplify_with_ranges` pass, which pretty much reimplements a best-effort basis version of this.

Now, in this new world order where we do have full control over all the simplifications that SymPy does, we can just use all the knowledge we have (guards and bounds on variables) to automatically perform these simplifications whenever possible. This will give us that, if we have a `PythonFloorDiv` in a expression, we know that we cannot simplify the numerator and denominator any further with the knowledge we have, as otherwise we would have done so at expression creation.

Having this shape_env allows us to perform that simplification you mention above at any point in time as, whenever we need to do it, we can just kindly ask `shape_env` whether the simplification is sound, and if it is, just perform it.